core/src/main/java/org/pentaho/di/core/plugins/PluginInterface.java

```diff
@@ -0,0 +1,215 @@
+   *         usually happens if a jar file is not installed correctly (class not found exceptions) etc.
+   */
+  String getErrorHelpFile();
+
+  /**
+   * @param errorHelpFile
+   *          the errorHelpFile to set
+   */
+  void setErrorHelpFile( String errorHelpFile );
+
+  URL getPluginDirectory();
+
+  /**
+   * @return the documentationUrl
+   */
+  String getDocumentationUrl();
+
+  /**
+   * @param documentationUrl
+   *          the documentationUrl to set
+   */
+  void setDocumentationUrl( String documentationUrl );
+
+  /**
+   * @return The cases URL of the plugin
+   */
+  String getCasesUrl();
+
+  /**
+   * @param casesUrl
+   *          the cases URL to set for this plugin
+   */
+  void setCasesUrl( String casesUrl );
+
+  /**
+   * @return the forum URL
+   */
+  String getForumUrl();
+
+  /**
+   * @param forumUrl
+   *          the forum URL to set
+   */
+  void setForumUrl( String forumUrl );
+
+  /**
+   * @return The group to which this class loader belongs.  
+   * Returns null if the plugin does not belong to a group (the default)
+   */
+  String getClassLoaderGroup();
+
+  void setSuggestion( String suggestion );
+
+  String getSuggestion();
+
+  /**
+   * @param group The group to which this class loader belongs.  
+   * Set to null if the plugin does not belong to a group (the default)
+   */
+  void setClassLoaderGroup( String group );
+
+  /**
+   * @param fragment A plugin interface to merge with
+   */
+  default void merge( PluginInterface fragment ) {
+    if ( fragment != null ) {
+      Optional.ofNullable( fragment.getClassMap() ).ifPresent( this.getClassMap()::putAll );
+      Optional.ofNullable( fragment.getImageFile() ).ifPresent( this::setImageFile );
+      Optional.ofNullable( fragment.getLibraries() ).ifPresent( this.getLibraries()::addAll );
+      Optional.ofNullable( fragment.getErrorHelpFile() ).ifPresent( this::setErrorHelpFile );
+      Optional.ofNullable( fragment.getDocumentationUrl() ).ifPresent( this::setDocumentationUrl );
+      Optional.ofNullable( fragment.getCasesUrl() ).ifPresent( this::setCasesUrl );
+      Optional.ofNullable( fragment.getForumUrl() ).ifPresent( this::setForumUrl );
+      Optional.ofNullable( fragment.getClassLoaderGroup() ).ifPresent( this::setClassLoaderGroup );
+    }
+  }
+
+  /**
+   * @return Plugin properties
+   * Returns a Properties object containing any properties in a plugin.properties file
+   * in the plugin directopry. The method will return null if the plugin.properties
+   * does not exist.
+   **/
+  default Properties getPluginProperties() throws KettleException {
+    String pluginPropsFilePath = getPluginDirectory().getPath() + "/plugin.properties";
+    File pluginPropsFile = new File( pluginPropsFilePath );
+    if ( !pluginPropsFile.exists() ) {
+      return null;
+    }
+    Properties pluginProps = new Properties();
+    try {
+      pluginProps.load( new FileInputStream( pluginPropsFile ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *FileInputStream*. Execution paths that do not contain closure statements were detected. Ensure the object returned by *FileInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).